src/_posts/2020-04-28-terraform-with-multiple-environments.md

```diff
@@ -0,0 +1,167 @@
+---
+layout: post
+asset-type: post
+name: terraform-with-multiple-environments
+title: Terraform With Multiple Environments
+date: 2020-04-22 10:00:00 +00:00
+author: Jorge Gueorguiev Garcia
+description: I have used a couple of different approaches to support multiple environments for Terraform deployments. We will investigate them here, and look at a third option.
+
+image:
+    src: /assets/custom/img/blog/railway.png
+    attribution:
+        text: Public domain photo from Pixabay
+        href: https://pixabay.com/photos/rails-soft-gleise-railway-3309912/
+tags:
+    - infrastructure
+    - devops
+    - terraform
+    - practices
+abstract: I have used a couple of different approaches to support multiple environments for Terraform deployments. We will investigate them here, and look at a third option.
+---
+
+# The setup
+
+I haven't written on my company blog for quite a while (been paying attention to the podcasts and my personal one), but I thought it was time to add a blog post. Part of what I am going to say here is coloured by using [Terraform](https://www.terraform.io/) only on [AWS](https://aws.amazon.com/).
+
+The topic I have chosen is because of this post that I discovered recently: [Recommended Practices](https://www.terraform.io/docs/cloud/guides/recommended-practices/part1.html). On it [Hashicorp](https://www.hashicorp.com/) is recommending the use of workspaces to control different environments. Which is a change of recommendation from what I remember from February/March 2019.
+
+# The first way - Single account with workspaces
+
+Back at the beginning of 2019 I started using Terraform for an AWS project. The basis of the infrastructure I was doing was quite similar to a previous project that I worked on (both in [Clojure](https://clojure.org/)), but that project used extensively bash scripts to deal with the creation of an ELB environment. You don't want to use bash scripts for everything, it is painful, messy ... soul crushing. So I decided to learn and use Terraform. Terraform is awesome (not perfect, small niggling issues here and there). Because on that project I had only one AWS account (I know better now), I needed to find a way to deploy multiple environments into that single account. For that I used [workspaces](https://www.terraform.io/docs/state/workspaces.html). When you are using workspaces you use a single set of files and then you have two options for the data/variables that can change, either you use different tfvar files, one per environment, passing it as a parameter to the terraform call (e.g., `terraform plan -var-file=env.dev.tfvars`).
```

Comment:
```suggestion
Back at the beginning of 2019 I started using Terraform for an AWS project. The basis of the infrastructure I was doing was quite similar to a previous project that I worked on (both in [Clojure](https://clojure.org/)), but that project used extensively bash scripts to deal with the creation of an ELB environment. You don't want to use bash scripts for everything, it is painful, messy ... soul crushing. So I decided to learn and use Terraform. Terraform is awesome (not perfect, small niggling issues here and there). Because on that project I had only one AWS account (I know better now), I needed to find a way to deploy multiple environments into that single account. For that I used [workspaces](https://www.terraform.io/docs/state/workspaces.html). When you are using workspaces you use a single set of files and then you have two options for the data/variables that can change, either you use different **tfvar** files, one per environment, passing it as a parameter to the terraform call (e.g., `terraform plan -var-file=env.dev.tfvars`).
```